Kirill_Samsonov/2/lib/checker.rb

```diff
@@ -0,0 +1,81 @@
+# frozen_string_literal: true
+
+require 'net/http'
+require_relative './reader'
+require_relative './output'
+
+class Checker
+  def initialize(source, options)
+    @source = source
+    @options = options
+  end
+
+  def run
+    reader = Reader.new(@source, @options)
+    domains = reader.read
+    domains = filter_domains(domains)
+
+    summary = check_domains(domains) do |domain_check_result|
+      Output.print_line format_domain_result(domain_check_result)
+    end
+
+    Output.print_line format_summary(summary)
+  end
+
+  def check_domains(domains)
+    result = { total: 0, success: 0, failed: 0, errored: 0 }
+    domains.each do |domain|
+      # Didn't find out how to make destructuring in es6 style. This looks ugly
+      response, elapsed_time = check_domain(domain).values_at(:response, :elapsed_time)
+      next if skip_domain?(response)
+
+      code = response.code.to_i
+      success_status?(code) ? result[:success] += 1 : result[:failed] += 1
+      yield ({ domain: domain, status: response.code, latency: elapsed_time })
+    rescue StandardError => e
+      result[:errored] += 1
+      yield ({ domain: domain, error: true, message: e.message })
+    end
+    result[:total] = result[:success] + result[:failed] + result[:errored]
+    result
+  end
+
+  def check_domain(name)
+    start_time = Time.now.utc
+    response = Net::HTTP.get_response(name, '/')
+    elapsed_time = ((Time.now.utc - start_time) * 1000).round
+    { response: response, elapsed_time: elapsed_time }
+  end
+
+  private
+
+  def skip_domain?(response)
+    filtered_by_name = @options[:filter] && !response.body.include?(@options[:filter])
+    filtered_open_source = @options[:exclude_solutions] && response.body.include?('open source')
+    filtered_by_name || filtered_open_source
+  end
+
+  def filter_domains(domains)
+    return domains unless @options[:no_subdomains]
+
+    domains.reject { |domain| subdomain?(domain) }
+  end
+
+  def subdomain?(domain)
+    domain.count('.') > 1
+  end
+
+  def success_status?(code)
+    code >= 200 && code < 400
+  end
+
+  def format_domain_result(line)
```

Comment:
At first I did it in Output, but later I refactor and move format here because I thought Output should know nothing about the internal representation of line and should only print the line to the appropriate output like stdout >>.

As instance if I would replace Output with OutputNetwork, it send formatted domain result string to the network. So maybe it's should be separate Formatter class and all format logic would be collected there.

Reply:
Yes, it shouldn't be in the Output. I meant just a separate class. The best option that came to my mind is the following:

1. There is an `Output` class, which prints something to console, as you have now.
1. There is a `Formatter` class, which takes the checker result and returns the string.
1. There is a `Presenter` class, which takes `Checker` instance (or checker result), `Formatter` instance, `Output` instance. And this class is responsible to present results to the user.

In such a way, we can change formatter/output on the way, and print results to console/GUI/HTTP-response with any available format.

Reply:
And then we should rename `Output` to the `ConsoleView`, for example.

Reply:
And create interfaces, like `IFormatter`, `IView`. But we can't do it with ruby ☹️